v3/src/components/graph/utilities/dot-plot-utils.ts

```diff
@@ -0,0 +1,243 @@
+          minBinEdge = 0, numExtraPrimaryBands, primaryAttrID, primaryAxisScale, totalNumberOfBins = 0 } = props
+  const caseValue = dataset?.getNumeric(anID, primaryAttrID) ?? -1
+  const binNumber = determineBinForCase(caseValue, totalNumberOfBins, binWidth) ?? 0
+  const binMidpoint = ((minBinEdge + binNumber * binWidth) - binWidth / 2) / numExtraPrimaryBands
+  const primaryCoord = isBinned
+    ? primaryAxisScale(binMidpoint)
+    : primaryAxisScale(dataset?.getNumeric(anID, primaryAttrID) ?? -1) / numExtraPrimaryBands
+  const extraPrimaryValue = dataset?.getStrValue(anID, extraPrimaryAttrID)
+  const extraPrimaryCoord = extraPrimaryValue ? extraPrimaryAxisScale(extraPrimaryValue ?? "__main__") ?? 0 : 0
+  return { primaryCoord, extraPrimaryCoord }
+}
+
+/*
+ * Returns a point's coordinate on the secondary axis in a dot plot or binned dot plot. It takes into account the
+ * primary and secondary axis scales, the primary and secondary bandwidths, the number of extra secondary bands, the
+ * overlap between points, the point diameter, and the primaryIsBottom flag.
+ */
+export const computeSecondaryCoord = (props: IComputeSecondaryCoord) => {
+  const { baseCoord, extraSecondaryAxisScale, extraSecondaryBandwidth, extraSecondaryCat, indexInBin,
+          numExtraSecondaryBands, overlap, pointDiameter, primaryIsBottom, secondaryAxisExtent, secondaryAxisScale,
+          secondaryBandwidth, secondaryCat, secondarySign } = props
+  let catCoord = (
+    !!secondaryCat && secondaryCat !== "__main__"
+      ? secondaryAxisScale(secondaryCat) ?? 0
+      : 0
+    ) / numExtraSecondaryBands
+  let extraCoord = !!extraSecondaryCat && extraSecondaryCat !== "__main__"
+    ? (extraSecondaryAxisScale(extraSecondaryCat) ?? 0) : 0
+  if (primaryIsBottom) {
+    extraCoord = secondaryAxisExtent - extraSecondaryBandwidth - extraCoord
+    catCoord = extraSecondaryBandwidth - secondaryBandwidth - catCoord
+    return baseCoord - catCoord - extraCoord -
+      pointDiameter / 2 - indexInBin * (pointDiameter - overlap)
+  } else {
+    return baseCoord + extraCoord + secondarySign * (catCoord + pointDiameter / 2 +
+      indexInBin * (pointDiameter - overlap))
+  }
+}
+
+/*
+ * Returns bins, binMap and overlap values.
+ */
+export const computeBinPlacements = (props: IComputeBinPlacements) => {
+  const { dataConfig, dataset, extraPrimaryAttrID, extraSecondaryAttrID, layout, numExtraPrimaryBands,
+          pointDiameter, primaryAttrID, primaryAxisScale, primaryPlace, secondaryAttrID, secondaryBandwidth,
+          totalNumberOfBins, binWidth } = props
+  const primaryLength = layout.getAxisLength(primaryPlace) / numExtraPrimaryBands
+  let overlap = 0
+  const numBins = totalNumberOfBins ? totalNumberOfBins : Math.ceil(primaryLength / pointDiameter) + 1
+  const useBinWidth = binWidth ? binWidth : primaryLength / (numBins - 1)
+  const bins: Record<string, Record<string, Record<string, string[][]>>> = {}
+  const binMap: Record<string, BinMap> = {}
```

Comment:
It need not be for this PR, but it seems like a `DotPlotBins` class that managed all of the bins and had methods for manipulating them and/or modifying them might be a way of managing some of this complexity and possibly even maintaining useful state across renders so that everything doesn't have to be computed from scratch on every render.